### PR TITLE
Implement Middle Click support for Slimefun Items

### PR DESCRIPTION
**Pull request by svr333**
## Description
This PR will add support to retrieve SlimefunItems by middle clicking on them.

## Proposed changes
- Add a listener to check for middle clicks on blocks

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:65]: I feel like there will be more edge cases like the player head coming up in the future.
So it would be a good idea to move this into a `private boolean` method so it can easily be modified later.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:66]: This would only work for Tile entities though, for general id grabbing you would need to use `BlockStorage`

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Thanks for adding these comments, I think I understand the proposed changes a lot better now

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:65]:
```suggestion
            /* 
             * Before giving the item to the user, check if you can swap
             * to the item instead (user already has item in hotbar).
             * This is sometimes bypassed by the client itself (not fixable though).
             */
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:68]: Can also do `sfItem.isItem(...)` here which should be better and more optimized

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:54]:
```suggestion
            SlimefunItem sfItem = BlockStorage.check(b);
```

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Review by TheBusyBiscuit — DISMISSED** [review]
LGTM

**Comment by svr333** [comment]
@Sfiguz7 whenever you arent too busy with exames, would you mind reviewing
you were there when I was coding this, might be useful insight

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:95]:
```suggestion
         * On a middle click outside the user inventory, cursor will be set
         * to the actual block that is middle clicked, while currentItem will be AIR.
         *
         * This check is really weird due to the weird nature of this event's behaviour.
         * It checks if the block the player is looking at is of the same type as the cursor;
         * after this we can make sure that it is a middle click outside of the inventory
         * currentItem should also be air, otherwise it is not outside of the inventory
         */
        boolean isOutsideInventoryClick = e.getCursor().getType() == b.getType() && e.getCurrentItem().getType() == Material.AIR;

        /*
         * This is an edge case where the player is looking at a WALL_HEAD (eg. cargo)
         * and then the boolean above wont match because WALL_HEAD != PLAYER_HEAD.
         * This check makes up for that lack.
         */
```

**Review by WalshyDev — APPROVED** [review]
LGTM

**Comment by svr333** [comment]
according to sefi/var there have been some changes to middle click in 1.18 so Ill have to do some more testing before merge

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:79]: params are nonnull by default :eyes:

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]
Pretty much approving this but I'd rather want this to be build 1001.
Luckily, I found an extremely dangerous bug that needs to be fixed before I would ever be able to merge it ;)

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java
Comment (svr333) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/listeners/MiddleClickListener.java:79]: NOOOO MY BUILD 1000 :(((((((((((((((

**Review by TheBusyBiscuit — APPROVED** [review]
LGTM

**Review by WalshyDev — APPROVED** [review]
LGTM - Won't merge yet

**Review by Sfiguz7 — APPROVED** [review]

**Pull request merged**
